### PR TITLE
[SYCLomatic] Refine the AnalysisInfo.h/cpp

### PR DESCRIPTION
**Pull request by zhiweij1**
This PR moves a lot of class method definitions from header file to source file and reorders the definitions in source file according to the order of declarations.

Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 1]
```diff
@@ -4414,14 +7025,6 @@ void FreeQueriesInfo::emplaceExtraDecl() {
       FilePath, ExtraDeclLoc, 0, OS.str(), nullptr));
 }
 
-template <class F, class... Ts>
-std::string buildStringFromPrinter(F Func, Ts &&...Args) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp:4485]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 2]
```diff
@@ -368,7 +336,7 @@ class DpctFileInfo {
                                   std::shared_ptr<Obj> Object) {
     return getMap<Obj>().insert(std::make_pair(Offset, Object)).first->second;
   }
-  inline const clang::tooling::UnifiedPath &getFilePath() { return FilePath; }
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:372]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 3]
```diff
@@ -383,59 +351,25 @@ class DpctFileInfo {
...
-    HeaderInsertedBitMap[HeaderType::HT_Math] = B;
-  }
-
-  void setAlgorithmHeaderInserted(bool B = true) {
-    HeaderInsertedBitMap[HeaderType::HT_Algorithm] = B;
-  }
-
-  void setTimeHeaderInserted(bool B = true) {
-    HeaderInsertedBitMap[HeaderType::HT_Time] = B;
-  }
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 4]
```diff
@@ -486,100 +420,49 @@ class DpctFileInfo {
...
-      llvm::dbgs() << "[DpctFileInfo::getLineInfo] illegal line number "
-                   << LineNumber;
-      static SourceLineInfo InvalidLine;
-      return InvalidLine;
-    }
-    return Lines[--LineNumber];
-  }
-  StringRef getLineString(unsigned LineNumber) {
-    return getLineInfo(LineNumber).Line;
-  }
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 5]
```diff
@@ -486,100 +420,49 @@ class DpctFileInfo {
...
+  const SourceLineInfo &getLineInfo(unsigned LineNumber);
+  StringRef getLineString(unsigned LineNumber);
 
   // Get line number by offset
-  inline unsigned getLineNumber(unsigned Offset) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:512]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 6]
```diff
@@ -752,102 +614,32 @@ class DpctGlobalInfo {
...
-      return false;
-    }
-  }
-  inline static bool isInAnalysisScope(SourceLocation SL) {
-    return isInAnalysisScope(DpctGlobalInfo::getLocInfo(SL).first);
-  }
-
-  static bool isInAnalysisScope(clang::tooling::UnifiedPath FilePath) {
-    return isChildPath(AnalysisScope, FilePath);
-  }
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 7]
```diff
@@ -752,102 +614,32 @@ class DpctGlobalInfo {
...
+  static bool isExcluded(const clang::tooling::UnifiedPath &FilePath);
   // TODO: implement one of this for each source language.
-  inline static bool isInCudaPath(SourceLocation SL) {
-    return isInCudaPath(getSourceManager()
-                            .getFilename(getSourceManager().getExpansionLoc(SL))
-                            .str());
-  }
+  static bool isInCudaPath(SourceLocation SL);
   // TODO: implement one of this for each source language.
-  static bool isInCudaPath(clang::tooling::UnifiedPath FilePath) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:813]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 8]
```diff
@@ -752,102 +614,32 @@ class DpctGlobalInfo {
...
-  static const clang::tooling::UnifiedPath &getAnalysisScope() {
-    return AnalysisScope;
-  }
-  static inline void addChangeExtensions(const std::string &Extension) {
-    assert(!Extension.empty());
-    ChangeExtensions.insert(Extension);
-  }
-  static inline const std::unordered_set<std::string> &getChangeExtensions() {
-    return ChangeExtensions;
-  }
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 9]
```diff
@@ -752,102 +614,32 @@ class DpctGlobalInfo {
...
+  static const clang::tooling::UnifiedPath &getInRoot();
+  static void setOutRoot(const clang::tooling::UnifiedPath &OutRootPath);
+  static const clang::tooling::UnifiedPath &getOutRoot();
+  static void
+  setAnalysisScope(const clang::tooling::UnifiedPath &InputAnalysisScope);
+  static const clang::tooling::UnifiedPath &getAnalysisScope();
+  static void addChangeExtensions(const std::string &Extension);
+  static const std::unordered_set<std::string> &getChangeExtensions();
   // TODO: implement one of this for each source language.
-  static void setCudaPath(const clang::tooling::UnifiedPath &InputCudaPath) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:878]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 10]
```diff
@@ -752,102 +614,32 @@ class DpctGlobalInfo {
...
-    CudaPath = InputCudaPath;
-  }
+  static void setCudaPath(const clang::tooling::UnifiedPath &InputCudaPath);
   // TODO: implement one of this for each source language.
-  static const clang::tooling::UnifiedPath &getCudaPath() {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:882]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 11]
```diff
@@ -752,102 +614,32 @@ class DpctGlobalInfo {
...
-    return CudaPath;
-  }
+  static const clang::tooling::UnifiedPath &getCudaPath();
+  static const std::string getCudaVersion();
 
-  static const std::string getCudaVersion() {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:887]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 12]
```diff
@@ -986,192 +738,63 @@ class DpctGlobalInfo {
...
-                              "--in-root-exclude.\n");
-      }
-    }
-  }
-  inline static std::unordered_map<std::string, bool> getExcludePath() {
-    return ExcludePath;
-  }
-  inline static std::set<ExplicitNamespace> getExplicitNamespaceSet() {
-    return ExplicitNamespaceSet;
-  }
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: return

File: clang/lib/DPCT/AnalysisInfo.h [hunk 13]
```diff
@@ -1279,56 +902,20 @@ class DpctGlobalInfo {
...
-  static inline std::string getTypeName(QualType QT) {
-    return getTypeName(QT, DpctGlobalInfo::getContext());
-  }
-  static inline std::string getUnqualifiedTypeName(QualType QT,
-                                                   const ASTContext &Context) {
-    return getTypeName(QT.getUnqualifiedType(), Context);
-  }
-  static inline std::string getUnqualifiedTypeName(QualType QT) {
-    return getUnqualifiedTypeName(QT, DpctGlobalInfo::getContext());
-  }
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 14]
```diff
@@ -1338,42 +925,16 @@ class DpctGlobalInfo {
...
-    setGetReplacedNamePtr(&getReplacedName);
-    llvm::raw_string_ostream OS(MigratedTypeStr);
-    clang::PrintingPolicy PP =
-        clang::PrintingPolicy(DpctGlobalInfo::getContext().getLangOpts());
-    QT.print(OS, PP);
-    OS.flush();
-    setGetReplacedNamePtr(nullptr);
-    return getFinalCastTypeNameStr(MigratedTypeStr);
-  }
-  static inline std::string getReplacedTypeName(QualType QT) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:1404]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 15]
```diff
@@ -1968,18 +1241,10 @@ class DpctGlobalInfo {
   DpctGlobalInfo &operator=(DpctGlobalInfo &&) = delete;
 
   // Wrapper of isInAnalysisScope for std::function usage.
-  static bool checkInAnalysisScope(SourceLocation SL) { return isInAnalysisScope(SL); }
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:2031]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 16]
```diff
@@ -2009,24 +1274,16 @@ class DpctGlobalInfo {
...
-  }
-  static inline SourceLocation getLocation(const FunctionDecl *FD) {
-    return FD->getBeginLoc();
-  }
-  static inline SourceLocation getLocation(const FieldDecl *FD) {
-    return FD->getLocation();
-  }
-  static inline SourceLocation getLocation(const CallExpr *CE) {
-    return CE->getEndLoc();
-  }
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 17]
```diff
@@ -2009,24 +1274,16 @@ class DpctGlobalInfo {
...
+  static SourceLocation getLocation(const VarDecl *VD);
+  static SourceLocation getLocation(const FunctionDecl *FD);
+  static SourceLocation getLocation(const FieldDecl *FD);
+  static SourceLocation getLocation(const CallExpr *CE);
   // The result will be also stored in KernelCallExpr.BeginLoc
-  static inline SourceLocation getLocation(const CUDAKernelCallExpr *CKC) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:2085]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 18]
```diff
@@ -4401,7 +2613,7 @@ class KernelCallExpr : public CallFunctionExpr {
   std::string Event;
   bool IsSync;
 
-  class {
+  class SubmitStmtsList_t {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: SubmitStmtsList

File: clang/lib/DPCT/AnalysisInfo.h [hunk 19]
```diff
@@ -4414,70 +2626,29 @@ class KernelCallExpr : public CallFunctionExpr {
...
-      Printer << List;
-      return Printer.newLine();
-    }
-  } SubmitStmtsList;
+                             StringRef Comments = "");
+  };
+  SubmitStmtsList_t SubmitStmtsList;
 
-  class {
+  class OuterStmts_t {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: OuterStmtsList

**Review by ziranzha — APPROVED** [review]

File: clang/lib/DPCT/AnalysisInfo.h [hunk 20]
```diff
@@ -752,102 +626,55 @@ class DpctGlobalInfo {
...
-        FM.getOptionalDirectoryRef(llvm::sys::path::parent_path(FilePathStr));
-    if (Dir) {
-      StringRef DirName = FM.getCanonicalName(*Dir);
-      StringRef FileName = llvm::sys::path::filename(FilePathStr);
-      llvm::sys::path::append(NoSymlinks, DirName, FileName);
-    }
-    return NoSymlinks.str().str();
-  }
-
-  inline static bool isInRoot(SourceLocation SL) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:780]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 21]
```diff
@@ -986,192 +773,67 @@ class DpctGlobalInfo {
...
   }
-  inline static const std::string &getGlobalDeviceName() {
+  static const std::string &getGlobalDeviceName() {
     const static std::string D = "dev_ct1";
     return D;
   }
-
   static std::string getStringForRegexReplacement(StringRef);
-
-  inline static void setCodeFormatStyle(const clang::format::FormatStyle &Style) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:1212]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 22]
```diff
@@ -986,192 +773,67 @@ class DpctGlobalInfo {
...
-    CodeFormatStyle = Style;
-  }
-  inline static clang::format::FormatStyle getCodeFormatStyle() {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:1215]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 23]
```diff
@@ -1279,56 +941,26 @@ class DpctGlobalInfo {
   getLocInfo(const T *N, bool *IsInvalid = nullptr /* out */) {
     return getLocInfo(getLocation(N), IsInvalid);
   }
-
   static std::pair<clang::tooling::UnifiedPath, unsigned>
-  getLocInfo(const TypeLoc &TL, bool *IsInvalid = nullptr /*out*/) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:1327]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 24]
```diff
@@ -1447,211 +1014,53 @@ class DpctGlobalInfo {
       HostDeviceFuncLocInfo Info, unsigned ID);
   void postProcess();
   void cacheFileRepl(clang::tooling::UnifiedPath FilePath,
-                     std::shared_ptr<ExtReplacements> Repl) {
-    FileReplCache[FilePath] = Repl;
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 25]
```diff
@@ -1447,211 +1014,53 @@ class DpctGlobalInfo {
...
-    }
-  }
+  void emplaceReplacements(ReplTy &ReplSets /*out*/);
   std::shared_ptr<KernelCallExpr> buildLaunchKernelInfo(const CallExpr *);
-
   void insertCudaMalloc(const CallExpr *CE);
   void insertCublasAlloc(const CallExpr *CE);
   std::shared_ptr<CudaMallocInfo> findCudaMalloc(const Expr *CE);
-  void addReplacement(std::shared_ptr<ExtReplacement> Repl) {
-    insertFile(Repl->getFilePath().str())->addReplacement(Repl);
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:1515]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 26]
```diff
@@ -2413,191 +1672,54 @@ class MemVarInfo : public VarInfo {
...
-  }
-  void setUseHelperFuncFlag(bool Flag) {
-    UseHelperFuncFlag = Flag;
-  };
+                                  bool NeedCheckExtraConstQualifier = false);
+  void setUseHelperFuncFlag(bool Flag) { UseHelperFuncFlag = Flag; }
   bool isUseHelperFunc() { return UseHelperFuncFlag; }
 
 private:
-  bool isTreatPointerAsArray() {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:2622]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 27]
```diff
@@ -3112,10 +1967,7 @@ class TemplateArgumentInfo {
     return Arg->getSourceRange();
   }
 
-  void setArgStr(std::string &&Str) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 28]
```diff
@@ -3523,35 +2150,23 @@ class CallFunctionExpr {
...
+  bool hasWrittenTemplateArgs();
+  const std::string &getName() { return Name; }
 
   std::string getTemplateArguments(bool &IsNeedWarning,
                                    bool WrittenArgsOnly = true,
                                    bool WithScalarWrapped = false);
 
   virtual std::string getExtraArguments();
 
-  inline void setHasSideEffects(bool Val = true) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:3605]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 29]
```diff
@@ -3523,35 +2150,23 @@ class CallFunctionExpr {
...
-    CallGroupFunctionInControlFlow = Val;
-  }
-  inline bool hasSideEffects() const { return CallGroupFunctionInControlFlow; }
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:3608]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 30]
```diff
@@ -3569,61 +2184,27 @@ class CallFunctionExpr {
 protected:
   void setFuncInfo(std::shared_ptr<DeviceFunctionInfo>);
   std::string Name;
-  inline unsigned getBegin() { return BeginLoc; }
-  inline const clang::tooling::UnifiedPath &getFilePath() { return FilePath; }
+  unsigned getBegin() { return BeginLoc; }
+  const clang::tooling::UnifiedPath &getFilePath() { return FilePath; }
   void buildInfo();
   void buildCalleeInfo(const Expr *Callee);
-  void resizeTextureObjectList(size_t Size) { TextureObjectList.resize(Size); }
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:3639]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 31]
```diff
@@ -3645,49 +2226,26 @@ class CallFunctionExpr {
...
-  }
-
-  inline std::shared_ptr<DeviceFunctionInfo> getFuncInfo() const {
-    return FuncInfo;
-  }
+      const ParsedAttributes &Attrs, const TemplateArgumentListInfo &TAList);
+  std::shared_ptr<DeviceFunctionInfo> getFuncInfo() const { return FuncInfo; }
 
   virtual void emplaceReplacement();
-  static void reset() { FuncInfoMap.clear(); };
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:3753]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 32]
```diff
@@ -3795,27 +2344,17 @@ class DeviceFunctionDeclInModule : public DeviceFunctionDecl {
   void buildParameterInfo(const FunctionDecl *FD);
   void buildWrapperInfo(const FunctionDecl *FD);
   void buildCallInfo(const FunctionDecl *FD);
-  std::vector<std::pair<std::string, std::string>> &getParametersInfo() {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 33]
```diff
@@ -4070,129 +2493,15 @@ class KernelCallExpr : public CallFunctionExpr {
...
-        IsDoublePointer = S->containsVirtualPointer();
-      }
-      ArgString = Obj->getName();
-      IdString = ArgString + "_";
-      ArgSize = MapNames::KernelArgTypeSizeMap.at(KernelArgType::KAT_Texture);
-    }
-
-    inline const std::string &getArgString() const { return ArgString; }
-    inline const std::string &getTypeString() const { return TypeString; }
-    inline std::string getIdStringWithIndex() const {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:4253]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 34]
```diff
@@ -4070,129 +2493,15 @@ class KernelCallExpr : public CallFunctionExpr {
...
-      return buildString(IdString, "ct", Index);
-    }
-    inline std::string getIdStringWithSuffix(const std::string &Suffix) const {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 35]
```diff
@@ -4221,93 +2530,44 @@ class KernelCallExpr : public CallFunctionExpr {
...
-                            const CallExpr *);
+  static std::shared_ptr<KernelCallExpr> buildFromCudaLaunchKernel(
+      const std::pair<clang::tooling::UnifiedPath, unsigned> &LocInfo,
+      const CallExpr *);
   static std::shared_ptr<KernelCallExpr>
   buildForWrapper(clang::tooling::UnifiedPath, const FunctionDecl *,
                   std::shared_ptr<DeviceFunctionInfo>);
   unsigned int GridDim = 3;
   unsigned int BlockDim = 3;
-  void setEmitSizeofWarningFlag(bool Flag) { EmitSizeofWarning = Flag; }
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: keep

File: clang/lib/DPCT/AnalysisInfo.h [hunk 36]
```diff
@@ -4221,93 +2530,44 @@ class KernelCallExpr : public CallFunctionExpr {
...
-        bool Used = true;
-        if (auto *ArgDRE = dyn_cast<DeclRefExpr>(Arg->IgnoreImpCasts()))
-          Used = isArgUsedAsLvalueUntil(ArgDRE, CE);
-        const auto FD = CE->getDirectCallee();
-        ArgsInfo.emplace_back(FD ? FD->parameters()[Idx] : nullptr, Analysis,
-                              Arg, Used, Idx, this);
-      }
-    }
-  }
-  bool isDefaultStream() const {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:4359]: keep

**Comment by zhiweij1** [comment]
Failed cases are caused by the new added codepin option. Test case needs to be updated.

**Pull request merged**